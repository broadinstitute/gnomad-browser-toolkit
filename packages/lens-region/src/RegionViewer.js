import React, { PropTypes, Component } from 'react'
import injectTapEventPlugin from 'react-tap-event-plugin'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from '@lens/utils-genomic-coordinates'  // eslint-disable-line

import defaultStyles from './styles.css'

injectTapEventPlugin()

const {
  exonColor,
  paddingColor,
  masterExonThickness,
  masterPaddingThickness,
} = defaultStyles

class RegionViewer extends Component {

  static propTypes = {
    css: PropTypes.object,
    regions: PropTypes.array.isRequired,
    regionAttributes: PropTypes.object,
    padding: PropTypes.number.isRequired,
    exonSubset: PropTypes.array,
    onRegionClick: PropTypes.func,
  }

  static defaultProps = {
    css: defaultStyles,
    exonSubset: null,
    onRegionClick: () => {},
    regionAttributes: {
      CDS: {
        color: exonColor,
        thickness: masterExonThickness,
      },
      start_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      end_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      intron: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      default: {
        color: 'grey',
        thickness: masterPaddingThickness,
      },
    },
  }

  state = {
    leftPanelWidth: 100,
    rightPanelWidth: 50,
    featuresToDisplay: ['CDS'],
    ready: false,
  }

  setWidth = (event, newValue) => {
    const newWidth = 800 * newValue
    this.setState({ width: newWidth })
  }

  setLeftPanelWidth = (event, newValue) => {
    const leftPanelWidth = Math.floor(400 * newValue)
    this.setState({ leftPanelWidth })
  }

  renderChildren = (childProps) => {
    return React.Children.map(this.props.children, (child) => {
      if (child) {
        return React.cloneElement(child, childProps)
      }
    })
  }

  render() {
    const { css } = this.props
    const { featuresToDisplay, leftPanelWidth } = this.state
    const {
      regions,
      regionAttributes,
      width,
      exonSubset,
      padding,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions,
      exonSubset,
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      regionAttributes,
      padding,
    }

    return (
      <div className={css.regionViewer}>
        {/*<p>Exon padding {padding.toPrecision(3)} bp</p>*/}
        <div style={{ width: width + leftPanelWidth }} className={css.regionArea}>
          {this.renderChildren(childProps)}
        </div>
      </div>
    )
  }
}
export default RegionViewer
