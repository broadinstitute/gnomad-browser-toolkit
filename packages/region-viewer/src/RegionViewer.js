import PropTypes from 'prop-types'
import React, { Component } from 'react'
import styled from 'styled-components'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from './coordinates'

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
  width: ${props => props.width}px;
  margin: 0 auto 10px;
  font-size: 12px;
`

const exonColor = '#212121'
const paddingColor = '#BDBDBD'
const masterExonThickness = '25px'
const masterPaddingThickness = '5px'

const defaultRegionAttributes = {
  CDS: {
    color: exonColor,
    thickness: masterExonThickness,
  },
  start_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  end_pad: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  intron: {
    color: paddingColor,
    thickness: masterPaddingThickness,
  },
  default: {
    color: 'grey',
    thickness: masterPaddingThickness,
  },
}

export class RegionViewer extends Component {
  static propTypes = {
    children: PropTypes.node,
    featuresToDisplay: PropTypes.arrayOf(PropTypes.string),
    leftPanelWidth: PropTypes.number,
    padding: PropTypes.number.isRequired,
    regionAttributes: PropTypes.objectOf(
      PropTypes.shape({
        color: PropTypes.string.isRequired,
        thickness: PropTypes.string.isRequired,
      })
    ),
    regions: PropTypes.arrayOf(
      PropTypes.shape({
        feature_type: PropTypes.string.isRequired,
        start: PropTypes.number.isRequired,
        stop: PropTypes.number.isRequired,
      })
    ).isRequired,
    rightPanelWidth: PropTypes.number,
    width: PropTypes.number.isRequired,
  }

  static defaultProps = {
    children: undefined,
    featuresToDisplay: ['CDS'],
    leftPanelWidth: 100,
    regionAttributes: defaultRegionAttributes,
    rightPanelWidth: 160,
  }

  renderChildren(childProps) {
    const { children } = this.props
    return React.Children.map(
      children,
      child => (child ? React.cloneElement(child, childProps) : null)
    )
  }

  render() {
    const {
      featuresToDisplay,
      regions,
      regionAttributes,
      width,
      padding,
      leftPanelWidth,
      rightPanelWidth,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      rightPanelWidth,
    }

    return (
      <RegionViewerWrapper width={width + leftPanelWidth + rightPanelWidth}>
        {this.renderChildren(childProps)}
      </RegionViewerWrapper>
    )
  }
}
