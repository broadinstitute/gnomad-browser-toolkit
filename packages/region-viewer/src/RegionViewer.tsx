import React from 'react'
import styled from 'styled-components'

import { mergeOverlappingRegions, regionViewerScale, Region, ScalePosition } from './coordinates'

export type RegionViewerContextValue = {
  centerPanelWidth: number
  isPositionDefined: (pos: number) => boolean
  leftPanelWidth: number
  regions: Region[]
  rightPanelWidth: number
  scalePosition: ScalePosition
}

const defaultRegionViewerContextValue: RegionViewerContextValue = {
  centerPanelWidth: 1440,
  isPositionDefined: () => false,
  leftPanelWidth: 100,
  regions: [],
  rightPanelWidth: 160,
  scalePosition: regionViewerScale([], [0, 1440])
}

export const RegionViewerContext = React.createContext<RegionViewerContextValue>(
  defaultRegionViewerContextValue
)

type Props = {
  children?: React.ReactNode
  leftPanelWidth: number
  regions: Region[]
  rightPanelWidth: number
  width: number
}

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

export const RegionViewer = ({
  children,
  leftPanelWidth = 100,
  regions,
  rightPanelWidth = 160,
  width
}: Props) => {
  const mergedRegions = mergeOverlappingRegions([...regions].sort((r1, r2) => r1.start - r2.start))

  const isPositionDefined = (pos: number) =>
    mergedRegions.some(region => region.start <= pos && pos <= region.stop)

  const centerPanelWidth = width - (leftPanelWidth + rightPanelWidth)
  const scalePosition = regionViewerScale(mergedRegions, [0, centerPanelWidth])

  const childProps = {
    centerPanelWidth,
    isPositionDefined,
    leftPanelWidth,
    regions,
    rightPanelWidth,
    scalePosition
  }

  return (
    <RegionViewerWrapper style={{ width }}>
      <RegionViewerContext.Provider value={childProps}>{children}</RegionViewerContext.Provider>
    </RegionViewerWrapper>
  )
}
