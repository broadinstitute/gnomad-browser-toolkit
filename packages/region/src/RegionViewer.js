import React, { Component } from 'react'
import PropTypes from 'prop-types'
import styled from 'styled-components'

import {
  calculateOffsetRegions,
  calculatePositionOffset,
  invertPositionOffset,
  calculateXScale,
} from '@broad/utilities/src/coordinates'  // eslint-disable-line

const RegionViewerWrapper = styled.div`
  display: flex;
  flex-direction: column;
  align-items: center;
  padding-left: 10px;
  padding-bottom: 10px;
`

const RegionArea = styled.div`
  flex-direction: column;
`

const exonColor = '#212121'
const paddingColor = '#BDBDBD'
const masterExonThickness = '25px'
const masterPaddingThickness = '5px'

class RegionViewer extends Component {
  static propTypes = {
    regions: PropTypes.array.isRequired,
    regionAttributes: PropTypes.object,
    padding: PropTypes.number.isRequired,
    width: PropTypes.number.isRequired,
    leftPanelWidth: PropTypes.number.isRequired,
    exonSubset: PropTypes.array,
    broadcast: PropTypes.func,
    featuresToDisplay: PropTypes.array,
  }

  static defaultProps = {
    exonSubset: null,
    leftPanelWidth: 100,
    onRegionClick: () => {},
    broadcast: () => {},
    featuresToDisplay: ['CDS'],
    regionAttributes: {
      CDS: {
        color: exonColor,
        thickness: masterExonThickness,
      },
      start_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      end_pad: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      intron: {
        color: paddingColor,
        thickness: masterPaddingThickness,
      },
      default: {
        color: 'grey',
        thickness: masterPaddingThickness,
      },
    },
  }

  state = {
    rightPanelWidth: 50,
    ready: false,
  }

  componentWillMount () {
    this.broadcastOffsetRegions()
  }

  componentDidUpdate () {
    this.broadcastOffsetRegions()
  }

  setWidth = (event, newValue) => {
    const newWidth = 800 * newValue
    this.setState({ width: newWidth })
  }

  setLeftPanelWidth = (event, newValue) => {
    const leftPanelWidth = Math.floor(400 * newValue)
    this.setState({ leftPanelWidth })
  }

  broadcastOffsetRegions = () => {
    if (this.props.regions) {
      const offsetRegions = calculateOffsetRegions(
        this.props.featuresToDisplay,
        this.props.regionAttributes,
        this.props.padding,
        this.props.regions,
        this.props.exonSubset
      )
      this.props.broadcast({ offsetRegions })
    }
  }

  renderChildren = (childProps) => {
    // eslint-disable-next-line
    return React.Children.map(this.props.children, (child) => {
      if (child) {
        return React.cloneElement(child, childProps)
      }
    })
  }

  render() {
    const { featuresToDisplay } = this.props
    const {
      regions,
      regionAttributes,
      width,
      exonSubset,
      padding,
      leftPanelWidth,
    } = this.props

    const offsetRegions = calculateOffsetRegions(
      featuresToDisplay,
      regionAttributes,
      padding,
      regions,
      exonSubset
    )

    const positionOffset = calculatePositionOffset(offsetRegions)
    const xScale = calculateXScale(width, offsetRegions)
    const invertOffset = invertPositionOffset(offsetRegions, xScale)

    const childProps = {
      leftPanelWidth,
      positionOffset,
      invertOffset,
      xScale,
      width,
      offsetRegions,
      regionAttributes,
      padding,
    }

    return (
      <RegionViewerWrapper>
        <RegionArea width={width + leftPanelWidth}>
          {this.renderChildren(childProps)}
        </RegionArea>
      </RegionViewerWrapper>
    )
  }
}
export default RegionViewer
