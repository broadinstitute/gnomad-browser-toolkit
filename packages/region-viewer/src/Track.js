import PropTypes from 'prop-types'
import React from 'react'
import styled from 'styled-components'

import { RegionViewerContext } from './RegionViewer'

const OuterWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

const InnerWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: stretch;
`

const TopPanel = styled.div`
  display: flex;
  width: ${props => props.width}px;
  margin-right: ${props => props.marginRight}px;
  margin-left: ${props => props.marginLeft}px;
`

const SidePanel = styled.div`
  display: flex;
  flex-shrink: 0;
  flex-direction: column;
  width: ${props => props.width}px;
`

const CenterPanel = styled.div`
  display: flex;
  flex-shrink: 0;
  flex-direction: column;
  width: ${props => props.width}px;
`

const TitlePanel = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
`

// eslint-disable-next-line react/prop-types
const defaultRenderLeftPanel = ({ title = '' }) => (
  <TitlePanel>
    {title.split('\n').map(s => (
      <span key={s}>{s}</span>
    ))}
  </TitlePanel>
)

export const Track = ({ children, renderLeftPanel, renderRightPanel, renderTopPanel, ...rest }) => (
  <RegionViewerContext.Consumer>
    {({
      centerPanelWidth,
      isPositionDefined,
      leftPanelWidth,
      offsetRegions,
      rightPanelWidth,
      scalePosition,
    }) => (
      <OuterWrapper>
        {renderTopPanel && (
          <TopPanel
            marginLeft={leftPanelWidth}
            marginRight={rightPanelWidth}
            width={centerPanelWidth}
          >
            {renderTopPanel({ ...rest, width: centerPanelWidth })}
          </TopPanel>
        )}
        <InnerWrapper>
          <SidePanel width={leftPanelWidth}>
            {renderLeftPanel && renderLeftPanel({ ...rest, width: leftPanelWidth })}
          </SidePanel>
          <CenterPanel width={centerPanelWidth}>
            {children({
              ...rest,
              isPositionDefined,
              leftPanelWidth,
              offsetRegions,
              rightPanelWidth,
              scalePosition,
              width: centerPanelWidth,
            })}
          </CenterPanel>
          {renderRightPanel && (
            <SidePanel width={rightPanelWidth}>
              {renderRightPanel({ ...rest, width: rightPanelWidth })}
            </SidePanel>
          )}
        </InnerWrapper>
      </OuterWrapper>
    )}
  </RegionViewerContext.Consumer>
)

Track.propTypes = {
  children: PropTypes.func.isRequired,
  renderLeftPanel: PropTypes.func,
  renderRightPanel: PropTypes.func,
  renderTopPanel: PropTypes.func,
}

Track.defaultProps = {
  renderLeftPanel: defaultRenderLeftPanel,
  renderRightPanel: null,
  renderTopPanel: null,
}
