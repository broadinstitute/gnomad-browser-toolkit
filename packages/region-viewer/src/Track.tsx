import React from 'react'
import styled from 'styled-components'

import { RegionViewerContext, RegionViewerContextValue } from './RegionViewer'

const OuterWrapper = styled.div`
  display: flex;
  flex-direction: column;
`

const InnerWrapper = styled.div`
  display: flex;
  flex-direction: row;
  align-items: stretch;
`

const TopPanel = styled.div`
  display: flex;
`

const SidePanel = styled.div`
  display: flex;
  flex-shrink: 0;
  flex-direction: column;
`

const CenterPanel = styled.div`
  display: flex;
  flex-shrink: 0;
  flex-direction: column;
`

const TitlePanel = styled.div`
  display: flex;
  flex-direction: column;
  justify-content: center;
  height: 100%;
`

export interface SidePanelProps {
  width: number
}

type _CenterPanelProps = Omit<RegionViewerContextValue, 'centerPanelWidth'> & { width: number }

export interface CenterPanelProps extends _CenterPanelProps {}

type Props = {
  renderLeftPanel?: React.FC<SidePanelProps>
  renderRightPanel?: React.FC<SidePanelProps>
  renderTopPanel?: React.FC<SidePanelProps>
  children: (props: CenterPanelProps) => React.JSX.Element
  [x: string]: unknown
}

const defaultRenderLeftPanel = ({ title = '' }: SidePanelProps & { title: string }) => (
  <TitlePanel>
    {title.split('\n').map(s => (
      <span key={s}>{s}</span>
    ))}
  </TitlePanel>
)

export const Track = ({
  children,
  renderLeftPanel = defaultRenderLeftPanel,
  renderRightPanel,
  renderTopPanel,
  ...rest
}: Props) => (
  <RegionViewerContext.Consumer>
    {({
      centerPanelWidth,
      isPositionDefined,
      leftPanelWidth,
      regions,
      rightPanelWidth,
      scalePosition,
    }) => {
      return (
        <OuterWrapper>
          {renderTopPanel && (
            <TopPanel
              style={{
                width: centerPanelWidth,
                marginLeft: leftPanelWidth,
                marginRight: rightPanelWidth,
              }}
            >
              {renderTopPanel({ ...rest, width: centerPanelWidth })}
            </TopPanel>
          )}
          <InnerWrapper>
            <SidePanel style={{ width: leftPanelWidth }}>
              {renderLeftPanel && renderLeftPanel({ ...rest, width: leftPanelWidth })}
            </SidePanel>
            <CenterPanel style={{ width: centerPanelWidth }}>
              {children({
                ...rest,
                width: centerPanelWidth,
                isPositionDefined,
                leftPanelWidth,
                regions,
                rightPanelWidth,
                scalePosition,
              })}
            </CenterPanel>
            {renderRightPanel && (
              <SidePanel style={{ width: rightPanelWidth }}>
                {renderRightPanel({ ...rest, width: rightPanelWidth })}
              </SidePanel>
            )}
          </InnerWrapper>
        </OuterWrapper>
      )
    }}
  </RegionViewerContext.Consumer>
)
