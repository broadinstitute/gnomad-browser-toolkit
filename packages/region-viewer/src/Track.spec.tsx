import React from 'react'
import { expect, test, describe } from '@jest/globals'
import { createRoot } from 'react-dom/client'
import 'jest-styled-components'

import { RegionViewer } from './RegionViewer'
import { CenterPanelProps, SidePanelProps, Track } from './Track'

const flushPromises = () => new Promise(resolve => setTimeout(resolve, 0))

describe('Track', () => {
  const TopPanelContent = ({ width, topLabel }: { width: number; topLabel: string }) => (
    <div style={{ width }}>Top panel says "{topLabel}"</div>
  )

  const LeftPanelContent = ({ width, leftLabel }: SidePanelProps & { leftLabel: string }) => (
    <div style={{ width }}>Left panel says "{leftLabel}"</div>
  )

  const RightPanelContent = ({ width, rightLabel }: SidePanelProps & { rightLabel: string }) => (
    <div style={{ width }}>Right panel says "{rightLabel}"</div>
  )

  const CenterPanelContent = ({
    leftPanelWidth,
    regions,
    rightPanelWidth,
    scalePosition,
    width,
    centerLabel
  }: CenterPanelProps & { centerLabel: string }) => (
    <div style={{ width }}>
      Greetings from main body of the track. "{centerLabel}"
      <ul>
        <li>leftPanelWidth: {leftPanelWidth}</li>
        <li>Regions: {JSON.stringify(regions)}</li>
        <li>rightPanelWidth: {rightPanelWidth}</li>
        <li>scalePosition: {scalePosition(150)}</li>
      </ul>
    </div>
  )

  test('has no unexpected changes', async () => {
    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer
        leftPanelWidth={200}
        rightPanelWidth={110}
        regions={[{ start: 123, stop: 456 }]}
        width={1200}
      >
        <Track
          renderLeftPanel={LeftPanelContent}
          renderRightPanel={RightPanelContent}
          renderTopPanel={TopPanelContent}
          leftLabel="I am on the left"
          rightLabel="I am on the right"
          centerLabel="Right here in the center"
          topLabel="Top label as in a heading or such"
        >
          {CenterPanelContent}
        </Track>
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })

  test('can omit any side panel', async () => {
    const container = document.createElement('div')
    const root = createRoot(container)
    root.render(
      <RegionViewer
        leftPanelWidth={200}
        rightPanelWidth={110}
        regions={[{ start: 123, stop: 456 }]}
        width={1200}
      >
        <Track
          renderLeftPanel={undefined}
          renderRightPanel={RightPanelContent}
          renderTopPanel={TopPanelContent}
          leftLabel="I am on the left"
          rightLabel="I am on the right"
          centerLabel="Right here in the center"
          topLabel="Top label as in a heading or such"
        >
          {CenterPanelContent}
        </Track>
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()

    root.render(
      <RegionViewer
        leftPanelWidth={200}
        rightPanelWidth={110}
        regions={[{ start: 123, stop: 456 }]}
        width={1200}
      >
        <Track
          renderLeftPanel={LeftPanelContent}
          renderRightPanel={RightPanelContent}
          renderTopPanel={undefined}
          leftLabel="I am on the left"
          rightLabel="I am on the right"
          centerLabel="Right here in the center"
          topLabel="Top label as in a heading or such"
        >
          {CenterPanelContent}
        </Track>
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()

    root.render(
      <RegionViewer
        leftPanelWidth={200}
        rightPanelWidth={110}
        regions={[{ start: 123, stop: 456 }]}
        width={1200}
      >
        <Track
          renderLeftPanel={LeftPanelContent}
          renderRightPanel={undefined}
          renderTopPanel={TopPanelContent}
          leftLabel="I am on the left"
          rightLabel="I am on the right"
          centerLabel="Right here in the center"
          topLabel="Top label as in a heading or such"
        >
          {CenterPanelContent}
        </Track>
      </RegionViewer>
    )
    await flushPromises()
    expect(container).toMatchSnapshot()
  })
})
