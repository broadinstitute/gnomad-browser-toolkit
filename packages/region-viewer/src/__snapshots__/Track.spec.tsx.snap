// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Track can omit any side panel 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c6 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
  height: 100%;
}

<div
    class="c0"
    style="width: 1200px;"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        style="width: 890px; margin-left: 200px; margin-right: 110px;"
      >
        <div
          style="width: 890px;"
        >
          Top panel says "Top label as in a heading or such"
        </div>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
          style="width: 200px;"
        >
          <div
            class="c5"
          >
            <span />
          </div>
        </div>
        <div
          class="c6"
          style="width: 890px;"
        >
          <div
            style="width: 890px;"
          >
            Greetings from main body of the track. "Right here in the center"
            <ul>
              <li>
                leftPanelWidth: 200
              </li>
              <li>
                Regions: [{"start":123,"stop":456}]
              </li>
              <li>
                rightPanelWidth: 110
              </li>
              <li>
                scalePosition: 71.94610778443113
              </li>
            </ul>
          </div>
        </div>
        <div
          class="c4"
          style="width: 110px;"
        >
          <div
            style="width: 110px;"
          >
            Right panel says "I am on the right"
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Track can omit any side panel 2`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
    style="width: 1200px;"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
      >
        <div
          class="c3"
          style="width: 200px;"
        >
          <div
            style="width: 200px;"
          >
            Left panel says "I am on the left"
          </div>
        </div>
        <div
          class="c4"
          style="width: 890px;"
        >
          <div
            style="width: 890px;"
          >
            Greetings from main body of the track. "Right here in the center"
            <ul>
              <li>
                leftPanelWidth: 200
              </li>
              <li>
                Regions: [{"start":123,"stop":456}]
              </li>
              <li>
                rightPanelWidth: 110
              </li>
              <li>
                scalePosition: 71.94610778443113
              </li>
            </ul>
          </div>
        </div>
        <div
          class="c3"
          style="width: 110px;"
        >
          <div
            style="width: 110px;"
          >
            Right panel says "I am on the right"
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Track can omit any side panel 3`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
    style="width: 1200px;"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        style="width: 890px; margin-left: 200px; margin-right: 110px;"
      >
        <div
          style="width: 890px;"
        >
          Top panel says "Top label as in a heading or such"
        </div>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
          style="width: 200px;"
        >
          <div
            style="width: 200px;"
          >
            Left panel says "I am on the left"
          </div>
        </div>
        <div
          class="c5"
          style="width: 890px;"
        >
          <div
            style="width: 890px;"
          >
            Greetings from main body of the track. "Right here in the center"
            <ul>
              <li>
                leftPanelWidth: 200
              </li>
              <li>
                Regions: [{"start":123,"stop":456}]
              </li>
              <li>
                rightPanelWidth: 110
              </li>
              <li>
                scalePosition: 71.94610778443113
              </li>
            </ul>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Track has no unexpected changes 1`] = `
<DocumentFragment>
  .c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c1 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c3 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: stretch;
  -webkit-box-align: stretch;
  -ms-flex-align: stretch;
  align-items: stretch;
}

.c2 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

.c4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

.c5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-shrink: 0;
  -ms-flex-negative: 0;
  flex-shrink: 0;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
}

<div
    class="c0"
    style="width: 1200px;"
  >
    <div
      class="c1"
    >
      <div
        class="c2"
        style="width: 890px; margin-left: 200px; margin-right: 110px;"
      >
        <div
          style="width: 890px;"
        >
          Top panel says "Top label as in a heading or such"
        </div>
      </div>
      <div
        class="c3"
      >
        <div
          class="c4"
          style="width: 200px;"
        >
          <div
            style="width: 200px;"
          >
            Left panel says "I am on the left"
          </div>
        </div>
        <div
          class="c5"
          style="width: 890px;"
        >
          <div
            style="width: 890px;"
          >
            Greetings from main body of the track. "Right here in the center"
            <ul>
              <li>
                leftPanelWidth: 200
              </li>
              <li>
                Regions: [{"start":123,"stop":456}]
              </li>
              <li>
                rightPanelWidth: 110
              </li>
              <li>
                scalePosition: 71.94610778443113
              </li>
            </ul>
          </div>
        </div>
        <div
          class="c4"
          style="width: 110px;"
        >
          <div
            style="width: 110px;"
          >
            Right panel says "I am on the right"
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
